api/client/options.proto

```diff
@@ -128,4 +128,6 @@ message CommandLineOptions {
   google.protobuf.BoolValue open_loop = 21;
   // See :option:`--jitter-uniform`
   google.protobuf.Duration jitter_uniform = 25 [(validate.rules).duration.gte.nanos = 1];
+  // See :option:`--backend-endpoint`
+  repeated google.protobuf.StringValue backend_endpoints = 27;
```

Comment:
Separately, I'm not sure that you're gaining any value here from using the protobuf.StringValue wrapper, rather than just using a string.  Maybe @oschaaf would have a better idea of whether that's true or not.

Reply:
To answer this, wrapping a scalar (like a string) in a message in proto3 adds a pointer. Whether the pointer is set or not can then be used as a signal of whether the field was explicitly set or not.

With a plain string in proto3, we can't recognize unset string from a string explicitly set to an empty string. Both would just read as an empty string. With the wrapper we can. For the case of an unset string, the outer message (StringValue) would have a nil pointer.

Practically we should avoid wrappers unless we need to distinguish between a set and an unset default value.

Reply:
@mum4k First -by no means am I a proto expert.

but this came up before, so let me share some rationale about why there is consistent wrapping used in this file. That is indeed done for the purpose of being able to discern between set and not set; but as of today for the options definition it is not concretely used. 

Having them wrapped is going to make things easy if we want to support hierarchical merging of options. Examples:
- have a base config on disk for nighthawk_client, and override from the CLI
- configure nighthawk_service with a base config, and override with incoming option values
This principle is used in the nighthawk-specific `nighthawk_test_server` configuration when performing overrides based on configuration delivered via incoming request headers. I think we can do this hierarchical merge with a single c++ call.

If we want to support this elsewhere, it will be easier with the wrapped values as we'll be able to discern between set and not set as you say. (Or actually, I think we close that door and make it impossible if we do not wrap?). I am curious what the penalty is that we're paying for the wrapping; if it's just being less efficient at the wire level / a performance penalty, I would be still inclined to go for wrapping as the consumers of this probably won't need to be able to do many transactions/second.. while we keep the door open for hierarchical configuration merging.

Also, perhaps less important, but I found the the consistent wrapping makes me have to think less as handling in code is uniform.

Reply:
Thanks for the added background Otto, I think we are on the same page in regards to why wrapping is required. Yes for determining if the value was set and also for the reason you provided - merging options. I think both of those are valid reasons to use the wrappers and while I am no proto expert myself, from what I know this won't add any significant penalty.

As a matter of fact, we are still using proto2 internally which uses pointers everywhere, so it follows the same pattern that allows to determine if a value was set. I don't see a reason to break away from this pattern and now we have good documented trail of why it is needed.